### PR TITLE
Add Contribution struct and instructions

### LINKED ISSUE #28: Implement Contribution Logic
Enable Backers to Contribute to Campaigns.

**Definition of Done:**
- [ ] Backers can pledge an amount to a campaign.
- [ ] Contributions are tracked via a Contribution account.
- [ ] Contributions are validated (e.g., within campaign deadline).

**Tasks:**
- [ ] Define the Contribution struct.
- [ ] Create an instruction for backers to contribute to a campaign.
- [ ] Validate contribution (check campaign deadline and status).
- [ ] Update campaign’s current_amount when a contribution is made.

**Code Example:**
```rs
#[account]
pub struct Contribution {
    pub campaign: Pubkey,             // Associated campaign account
    pub backer: Pubkey,               // Backer's wallet address
    pub amount: u64,                  // Amount contributed
    pub pledged_at: i64,              // Timestamp of the pledge
}

#[derive(Accounts)]
pub struct Contribute<'info> {
    #[account(mut)]
    pub campaign: Account<'info, Campaign>,
    #[account(init, payer = backer, space = 8 + Contribution::MAX_SIZE, seeds = [b"contribution", campaign.key().as_ref(), backer.key().as_ref()], bump)]
    pub contribution: Account<'info, Contribution>,
    #[account(mut)]
    pub backer: Signer<'info>,
    pub system_program: Program<'info, System>,
}

pub fn contribute(ctx: Context<Contribute>, amount: u64) -> Result<()> {
    let campaign = &mut ctx.accounts.campaign;
    require!(Clock::get()?.unix_timestamp < campaign.end_date, ErrorCode::CampaignExpired);
    require!(campaign.status == 0, ErrorCode::InvalidCampaignStatus);

    campaign.current_amount += amount;

    let contribution = &mut ctx.accounts.contribution;
    contribution.campaign = campaign.key();
    contribution.backer = *ctx.accounts.backer.key;
    contribution.amount = amount;
    contribution.pledged_at = Clock::get()?.unix_timestamp;

    Ok(())
}
```

### PR DESCRIPTION
**Pull request by loganrudd**
Adding contribution instruction logic and account struct, and creating test

### CODE REVIEW

File: anchor/programs/concert-x/src/lib.rs [hunk 1]
```diff
@@ -1,51 +1,78 @@
...
+    // * `title` - The title of the concert campaign
+    // * `short_description` - Brief description of the concert
+    // * `goal_amount` - Target funding amount in lamports
+    // * `start_date` - Unix timestamp for campaign start
+    // * `end_date` - Unix timestamp for campaign end
         ctx: Context<CreateConcert>,
         title: String,
         short_description: String,
-        goal_amount: u32,
+        goal_amount: u64,
```
Comment (loganrudd) [inline anchor/programs/concert-x/src/lib.rs]: change to u16

File: anchor/programs/concert-x/src/lib.rs [hunk 2]
```diff
@@ -1,51 +1,78 @@
...
+        ticket_price: u64,
```
Comment (loganrudd) [inline anchor/programs/concert-x/src/lib.rs]: change to u8

File: anchor/programs/concert-x/src/lib.rs [hunk 3]
```diff
@@ -1,51 +1,78 @@
...
         let concert = &mut ctx.accounts.concert;
-        concert.creator = ctx.accounts.initializer.key();
+        concert.pda = ctx.accounts.initializer.key();
         concert.title = title;
         concert.short_description = short_description;
         concert.goal_amount = goal_amount;
+        concert.ticket_price = ticket_price;
         concert.start_date = start_date;
         concert.end_date = end_date;
-        concert.max_token_supply = max_token_supply;
```
Comment (loganrudd) [inline anchor/programs/concert-x/src/lib.rs:39]: keep max_token_supply

File: anchor/programs/concert-x/src/lib.rs [hunk 4]
```diff
@@ -1,51 +1,78 @@
...
+        // * `ctx` - The context of the instruction
+        // * `amount` - The amount of lamports to contribute
+
+        //Require that campaign is active and contribution amount is greater than or equal ticket price
+        require!(
+            ctx.accounts.concert.status == 0,
+            ErrorCode::ConcertNotActive
+        );
+        require!(
+            amount >= ctx.accounts.concert.ticket_price,
```
Comment (loganrudd) [inline anchor/programs/concert-x/src/lib.rs:61]: amount = contribution_amount

File: anchor/programs/concert-x/src/lib.rs (hunk 3 above)
Comment (loganrudd) [reply anchor/programs/concert-x/src/lib.rs:39]: add list of contributors as the last field, and update in add_contribution logic

**Review by jdnichollsc — APPROVED** [review]
LGTM, nice job! <3

**Comment by jucoba** [comment]
Merge pr

**Pull request merged**